### DISCUSSION TITLE
Issue when device tries to connect via TLS MQTT with AWS Greengrass Core

### DISCUSSION THREAD

**Discussion by pipegreyback** [arendst/Tasmota#12599] [Q&A] [answered]
Hi there, I've been using Tasmota on some devices and I love this firmware so far.

Lately, I've been Implementing the following architecture:

![HEMS-New House](https://user-images.githubusercontent.com/13972706/124978878-54aa4980-e000-11eb-9ab0-755a5e5ade82.png)

The left side devices are mainly Sonoff Pow R2, running Tasmota 9.5.0.2. I've flashed the devices with a custom tasmota firmware, for enabling MQTT over TLS, and also USE_MQTT_AWS_IOT. I have followed the instructions described here https://tasmota.github.io/docs/AWS-IoT-cert/, which were enough for enabling the Sonoff to send data via TLS MQTT to AWS IoT directly. I had to upload a TLSKey 1 and TLSKey 2, I know there is a new scheme just using Client and Password, but I have not tried that one yet.

Currently, I am using a Raspberry Pi as a Greengrass Core, as you can see on the image, which acts as an MQTT Broker for local devices, and then re-directs the data to AWS IoT. I have kept the configuration on the sonoff, as they were for connecting to AWS IoT, and just changed the MQTT Host IP so now it uses the local IP of the Raspberry Pi running Greengrass Core. The issue's that Sonoff is not able to connect with the Raspberry Pi. I think there may be an issue with the certificates. Does anyone know how may I approach this issue? I have not seen information on Tasmota interacting with Greengrass Core devices, just with AWS IoT.

Here are some pictures:
![Screenshot_20210708_155126](https://user-images.githubusercontent.com/13972706/124982287-86bdaa80-e004-11eb-9603-a009ab261b52.png)

![Screenshot_20210708_155309](https://user-images.githubusercontent.com/13972706/124982440-b2409500-e004-11eb-90e3-1e643753b4eb.png)

**Comment by s-hadinger** [comment]
I never tried Green grass connection. The error 57 is "X.509 status: unknown critical extension in certificate."

Can you dump your certificate using openssl and paste it here?

**Reply by pipegreyback** [reply]
Thanks for answering @s-hadinger , here it's the content of the certificate:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            bc:56:6a:10:16:f2:4c:52:fd:3b:66:83:0d:d7:7c:33:e4:0e:c5:9f
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: OU = Amazon Web Services O=Amazon.com Inc. L=Seattle ST=Washington C=US
        Validity
            Not Before: Jul  4 01:42:06 2021 GMT
            Not After : Dec 31 23:59:59 2049 GMT
        Subject: C = EU
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:65:20:91:4d:2f:8a:12:95:0d:19:5d:76:a6:20:
                    1d:86:f2:44:db:ed:68:d9:e8:38:85:d4:32:42:89:
                    42:9a:ab:d0:c0:55:0e:e2:11:c3:9f:68:f0:9c:a7:
                    94:d8:c8:9c:d2:4e:9d:42:ad:4f:99:b3:85:25:af:
                    f1:8a:88:d2:19
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                keyid:19:FF:99:80:13:C4:08:3E:5A:B7:AC:1C:4F:C1:46:28:43:EC:9D:CA

            X509v3 Subject Key Identifier: 
                6E:3C:79:DE:BF:5E:CE:17:D0:7F:48:14:F3:14:07:29:52:3C:1E:A4
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature
    Signature Algorithm: sha256WithRSAEncryption
         14:59:ca:eb:6a:13:bd:07:89:86:49:e5:e0:26:70:54:e3:5c:
         36:20:85:f3:01:b5:9c:e2:8a:69:ed:64:55:3d:30:a3:73:3f:
         d0:8d:12:0d:39:e3:be:ab:4b:1f:16:92:9e:f2:7b:00:a5:c2:
         a7:89:02:61:b6:b4:1f:60:97:1d:90:75:48:d2:ea:95:95:3c:
         98:36:46:88:30:18:46:6b:bb:d1:44:a3:c3:7c:a6:64:94:ae:
         eb:58:28:80:1f:1f:07:3f:a3:b5:35:17:82:2f:90:d7:07:86:
         ae:d6:d9:e7:ea:66:de:ec:c1:85:fc:3d:9e:45:3b:87:c2:15:
         c5:c1:5b:7a:3c:d2:bd:56:f8:7f:48:e0:c8:78:b4:54:c5:f0:
         89:81:bd:b6:6b:96:c5:fb:2f:29:fa:56:4c:9c:01:30:af:2c:
         6e:aa:c0:1b:77:eb:6c:06:de:36:5a:a4:3c:d7:c2:5c:d0:4b:
         e2:d8:d4:1a:4a:ed:a0:8e:45:de:09:81:24:61:2b:4b:bf:38:
         76:6d:c0:5a:40:6b:45:02:2c:74:75:6f:da:37:e7:bd:37:f7:
         ce:a6:e1:39:61:c5:90:44:c7:65:a8:bb:94:da:5b:41:75:a0:
         2b:3c:92:67:ed:03:3c:76:52:e7:6f:bd:3c:22:fc:47:f2:f5:
         e6:0e:89:10
-----BEGIN CERTIFICATE-----
MIICfjCCAWagAwIBAgIVALxWahAW8kxS/Ttmgw3XfDPkDsWfMA0GCSqGSIb3DQEB
CwUAME0xSzBJBgNVBAsMQkFtYXpvbiBXZWIgU2VydmljZXMgTz1BbWF6b24uY29t
IEluYy4gTD1TZWF0dGxlIFNUPVdhc2hpbmd0b24gQz1VUzAeFw0yMTA3MDQwMTQy
MDZaFw00OTEyMzEyMzU5NTlaMA0xCzAJBgNVBAYTAkVVMFkwEwYHKoZIzj0CAQYI
KoZIzj0DAQcDQgAEZSCRTS+KEpUNGV12piAdhvJE2+1o2eg4hdQyQolCmqvQwFUO
4hHDn2jwnKeU2Mic0k6dQq1PmbOFJa/xiojSGaNgMF4wHwYDVR0jBBgwFoAUGf+Z
gBPECD5at6wcT8FGKEPsncowHQYDVR0OBBYEFG48ed6/Xs4X0H9IFPMUBylSPB6k
MAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMA0GCSqGSIb3DQEBCwUAA4IB
AQAUWcrrahO9B4mGSeXgJnBU41w2IIXzAbWc4opp7WRVPTCjcz/QjRINOeO+q0sf
FpKe8nsApcKniQJhtrQfYJcdkHVI0uqVlTyYNkaIMBhGa7vRRKPDfKZklK7rWCiA
Hx8HP6O1NReCL5DXB4au1tnn6mbe7MGF/D2eRTuHwhXFwVt6PNK9Vvh/SODIeLRU
xfCJgb22a5bF+y8p+lZMnAEwryxuqsAbd+tsBt42WqQ818Jc0Evi2NQaSu2gjkXe
CYEkYStLvzh2bcBaQGtFAix0dW/aN+e9N/fOpuE5YcWQRMdlqLuU2ltBdaArPJJn
7QM8dlLnb708IvxH8vXmDokQ
-----END CERTIFICATE-----
```

**Comment by s-hadinger** [comment]
Here is an excerpt from Bearssl doc: "Each certificate may contain a large number of extensions, with an open-ended format, that may impact the validation process in arbitrary ways. Each extension can be marked as “critical” or “not critical”: a critical extension is one that must not be ignored (thus, a system that does not recognise an extension can ignore it if it is non-critical, but it must reject the certificate if the extension is critical)"

Maybe `Basic Constraints: critical` is the issue

**Reply by s-hadinger** [reply]
I can try disabling the test for critical extensions but it's not straightforward. Let me take a look tomorrow

**Reply by pipegreyback** [reply]
Thanks for the help!

Let me know if there is any other information that I could provide.
I will take a look into the AWS policies for the Sonoff. I wonder if I should upload the certificate to the Raspberry Pi, or maybe the Greengrass Auth (the local component that runs in the Pi) is in charge of matching the sonoff keys with the AWS certificate in AWS IoT.

**Reply by s-hadinger** [reply]
Changing the policy won't solve it. I will try to patch BearSSL to relax its checks on critical attributes.

**Comment by s-hadinger** [comment, marked as answer]
I just merged the fix, please try from the latest development version.

See #12601

**Reply by pipegreyback** [reply]
I've just built from the development version, but I am still getting the 57 error from the tasmota console. It's weird that when I use the AWS endpoint, is able to connect with no issues, but when I change the MQTT host to my local Raspberry Pi IP, it throws that error.
For reference, this is the command I'm using in the console for the connection purpose:
`Backlog MqttHost <aws-info>.amazonaws.com; MqttPort 8883` and when I want to point it to my raspberry I use: `Backlog MqttHost 192.168.3.23; MqttPort 8883`
The rest it's the same, TLSKey 1 and TLSKey 2, obtained from AWS CloudFormation running the procedure as described in the archived guide on tasmota's website.

**Reply by s-hadinger** [reply]
You can't use an IP address for TLS, it needs to be a domain name matching the one of the certificate
